trio/_threads.py

```diff
@@ -70,6 +71,71 @@ class ThreadPlaceholder:
     name: str = attr.ib()
 
 
+# Types for the to_thread_run_sync message loop
+@attr.s(frozen=True, eq=False)
+class ThreadDone(Generic[RetT]):
+    result: outcome.Outcome[RetT] = attr.ib()
+
+
+@attr.s(frozen=True, eq=False)
+class Run(Generic[RetT]):
+    afn: Callable[..., Awaitable[RetT]] = attr.ib()
+    args: tuple[object, ...] = attr.ib()
+    context: contextvars.Context = attr.ib()
+    queue: stdlib_queue.SimpleQueue[outcome.Outcome[RetT]] = attr.ib(
+        init=False, factory=stdlib_queue.SimpleQueue
+    )
+
+    @disable_ki_protection
+    async def unprotected_afn(self) -> RetT:
+        coro = coroutine_or_error(self.afn, *self.args)
+        return await coro
+
+    async def run(self) -> None:
+        task = trio.lowlevel.current_task()
+        old_context = task.context
+        task.context = self.context.copy()
+        try:
+            await trio.lowlevel.cancel_shielded_checkpoint()
```

Comment:
Comment that this is to pick up the change in `context`. It's a shame that there's not a fast "these contexts definitely are equal, yes or no" operation to be able to bypass the extra checkpoint in the common case where the thread doesn't touch its context -- it would be easy to implement given the underlying representation of Context.